packages/contract/lib/PremiumCalculator.ts

```diff
@@ -0,0 +1,124 @@
+import { DualFundingTxFinalizer } from '@node-dlc/core';
+import { FundingInputV0 } from '@node-dlc/messaging';
+const p2wpkhSPK = Buffer.from(
+  '0014663117d27e78eb432505180654e603acb30e8a4a',
+  'hex',
+);
+
+export const buildDualFundingTxFinalizer = (
+  feeRate: number,
+): DualFundingTxFinalizer => {
+  if (feeRate <= 0) {
+    throw new Error('Fee rate must be greater than 0');
+  }
+
+  const fundingInput = new FundingInputV0();
+
+  fundingInput.maxWitnessLen = 108;
+  fundingInput.redeemScript = Buffer.from([]);
+
+  // assuming 1 utxo for funding
+  const finalizer = new DualFundingTxFinalizer(
+    [fundingInput],
+    p2wpkhSPK,
+    p2wpkhSPK,
+    [fundingInput],
+    p2wpkhSPK,
+    p2wpkhSPK,
+    BigInt(feeRate),
+  );
+
+  return finalizer;
+};
+
+type FeeConfig = {
+  serviceFees: number;
+  mmDiscount: number;
+};
+
+export class PremiumCalculator {
+  constructor(readonly feeConfig: FeeConfig) {}
+
+  /**
+   * Calcuates the premium received by the contract offeror after fees.
+   *
+   * If network fees exceed the fee collected by the service provider, no service fees are applied.
+   *
+   * @param premium
+   * @param feeRate
+   * @returns
+   */
+  calculatePostFeePremium = (premium: number, feeRate: number): number => {
+    const finalizer = buildDualFundingTxFinalizer(feeRate);
+
+    const acceptFees = finalizer.acceptFees,
+      offerFees = finalizer.offerFees;
+
+    const mmPremium =
+      BigInt(Math.floor(premium * (1 - this.feeConfig.mmDiscount))) -
+      acceptFees -
+      offerFees;
+    const maxPremium = BigInt(
+      Math.floor(
+        premium *
+          (1 - (this.feeConfig.mmDiscount + this.feeConfig.serviceFees)),
+      ),
+    );
+
+    const actualPremium = mmPremium > maxPremium ? maxPremium : mmPremium;
+    return Number(actualPremium);
+  };
+
+  /**
+   * Calcluate the premium received by the contract offeror before fees.
+   *
+   * @param premium
+   * @param feeRate
```

Comment:
would be nice to have a description for premium and feerate params

i.e. premium is in sats